qutebrowser/utils/psl.py

```diff
@@ -0,0 +1,145 @@
+#
+# You should have received a copy of the GNU General Public License
+# along with qutebrowser.  If not, see <https://www.gnu.org/licenses/>.
+
+import enum
+import os.path
+
+from PyQt5.QtCore import QUrl
+
+from qutebrowser.utils import standarddir, urlutils
+
+
+class PSLNotFound(Exception):
+    """Raised when no public_suffix_list.dat file is found.
+
+    Attributes:
+        paths: iterable of all tried paths.
+    """
+
+    def __init__(self, paths):
+        super().__init__(
+            "Couldn't find public_suffix_list.dat in any of the expected paths: "
+            "{}".format(', '.join(paths))
+        )
+
+
+class _Rule(enum.Enum):
+
+    NORMAL = enum.auto()
+    WILDCARD = enum.auto()
+    EXCEPTION = enum.auto()
+
+
+def _widened_hostnames(hostname):
+    """A generator for widening string hostnames.
+
+    Ex: a.c.foo -> [a.c.foo, c.foo, foo]"""
+    while hostname:
+        yield hostname
+        hostname = hostname.partition(".")[-1]
+
+
+class _SuffixList:
+
+    _rules = None
+
+    def _load_rules(self) -> None:
+        PSL_PATHS = (
+            # User-provided list in data directory (need to wait until standarddir is
+            # initialized)
+            os.path.join(standarddir.data(), 'public_suffix_list.dat'),
+            # Bundled list
+            os.path.join(__file__, '../public_suffix_list.dat'),
+            # Debian publicsuffix package
+            '/usr/share/publicsuffix/public_suffix_list.dat'
+        )
+        path = None
+        for psl_path in PSL_PATHS:
+            if os.path.exists(psl_path):
+                path = psl_path
+                break
+        if not path:
+            raise PSLNotFound(PSL_PATHS)
+
+        self._rules = {}
+        with open(path, 'r') as f:
+            for line in f:
+                line = line.strip()
+                if line.startswith('//') or not line:
+                    continue
+                if line.startswith('*.'):
+                    self._rules[line[2:]] = _Rule.WILDCARD
+                elif line.startswith('!'):
+                    self._rules[line[1:]] = _Rule.EXCEPTION
+                else:
+                    assert '*' not in line and '!' not in line, line
+                    self._rules[line] = _Rule.NORMAL
+
+    @property
+    def rules(self):
+        if self._rules is None:
+            self._load_rules()
+        return self._rules
+
+    def sld(self, url: QUrl) -> str:
+        urlutils.ensure_valid(url)
+        host = url.host()
+
+        if '.' not in host or self.rules.get(host) in (_Rule.NORMAL, _Rule.WILDCARD):
+            return None  # FIXME why?
+            # If the hostname doesn't have a dot it's either a tld, which doesn't have
+            # a sld, or it doesn't match any rule, in which case the prevailing rule is
+            # '*' according to https://publicsuffix.org/list/. If the hostname has a dot
+            # but matches a normal/wildcard rule, it's a tld.
```

Comment:
Leaving this comment for the sake of discussion, if still needed. I'll remove it afterwards.

Reply:
I think it'd be good to keep it in (and remove the "FIXME why?"). I don't think it's immediately obvious what the check does.